### PR TITLE
Add {secure:} opt to protocol.registerStandardSchemes

### PR DESCRIPTION
**Pull request by pfrazee**
This PR replaces https://github.com/electron/electron/pull/7671. It will:

 - [x] Add a second arg to `registerStandardSchemes`, which is an optional object where you can specify `{ secure: true }` to enable secure registration. Update `protocol.registerStandardSchemes` to register a scheme as secure on the browser side, and to propagate that registration to renderers.
 - [x] Add deprecation notices to `webFrame.registerURLSchemeAsSecure` and `webFrame.registerURLSchemeAsPrivileged`

Still in progress. I'm opening this PR to discuss the changes while I work.

### CODE REVIEW

**Comment by pfrazee** [comment]
I'll need some help on this -

I was having difficulty writing the tests, so I decided to run a demo app and just check the DevTools. I looked at the Security tab and got:

screen shot 2016-11-11 at 1 13 38 pm

So, for reference, I went back to my custom build of electron (in which I do have https://github.com/electron/electron/pull/7671 merged) and added `webFrame.registerURLSchemeAsPrivileged` to the demo. However, the devtools security tab still shows Insecure.

So, I'm wondering if even the current master, and my custom build, are missing a step for correctly registering a protocol as secure. Any thoughts?

**Comment by deepak1556** [comment]
The devtools panel security info is not standard, currently only valid HTTPS schemes are considered secure https://cs.chromium.org/chromium/src/components/security_state/security_state_model.h?l=38-66 . 

You can test if origin is registered as secure by trying to use cachestorage api, the renderer will terminate if the origin is not secure.

**Comment by pfrazee** [comment]
Ah, very good, thank you

**Comment by pfrazee** [comment]
How should the deprecations be handled? Do I remove the code and put a deprecation notice on the docs?

**Comment by MarshallOfSound** [comment]
@pfrazee Someone can confirm this ( @kevinsawicki ) but I believe the way we currently do it is leave the deprecated method in.  Add a console warning that it is deprecated.  Remove it from the docs.  And add a `// TODO` in the code to remove it entirely in the next major bump.

**Comment by pfrazee** [comment]
Ok, this is RFR

File: atom/renderer/api/atom_api_web_frame.cc [hunk 1]
```diff
@@ -129,9 +131,11 @@ void WebFrame::SetSpellCheckProvider(mate::Arguments* args,
 }
 
 void WebFrame::RegisterURLSchemeAsSecure(const std::string& scheme) {
+  // TODO(pfrazee): Remove 2.0
   // Register scheme to secure list (https, wss, data).
   blink::WebSecurityPolicy::registerURLSchemeAsSecure(
       blink::WebString::fromUTF8(scheme));
+  std::cout << "webFrame.registerURLSchemeAsSecure is deprecated. Use protocol.registerStandardSchemes with {secure:true}" << std::endl;
```
Comment (deepak1556) [inline atom/renderer/api/atom_api_web_frame.cc]: You can move the deprecations to `lib/renderer/web-frame.js`

File: docs/tutorial/planned-breaking-changes.md [hunk 2]
```diff
@@ -86,6 +86,20 @@ webContents.openDevTools({detach: true})
 webContents.openDevTools({mode: 'detach'})
 ```
 
+## `webContents`
+
+```js
+// Deprecated
+webFrame.registerURLSchemeAsSecure('app')
+// Replace with
+protocol.registerStandardSchemes('app', {secure: true})
```
Comment (deepak1556) [inline docs/tutorial/planned-breaking-changes.md]: `protocol.registerStandardSchemes` accepts only array of scheme(s).

File: atom/renderer/api/atom_api_web_frame.cc (hunk 1 above)
Comment (pfrazee) [inline atom/renderer/api/atom_api_web_frame.cc]: So, wrap the methods there so that they log?

**Comment by kevinsawicki** [comment]
> Add deprecation notices to 

We haven't really formally deprecated (meaning adding logged messages) any 2.0 APIs yet.

I think it is okay for now to just add a `TODO` comment like [this one](https://github.com/electron/electron/blob/0ef6d4631d2c18c41a09c7a763c212c825f8975f/atom/common/api/atom_bindings.cc#L118) and add an entry to [`planned-breaking-changes.md`](https://github.com/electron/electron/blob/master/docs/tutorial/planned-breaking-changes.md).

Then when we are closer to having a 2.0 API plan, we can add deprecation logging to all the needed areas.

**Comment by pfrazee** [comment]
Ok, I fixed the documentation. That should cover it.

File: atom/renderer/api/atom_api_web_frame.cc (hunk 1 above)
Comment (kevinsawicki) [inline atom/renderer/api/atom_api_web_frame.cc]: @pfrazee can this be removed and just documented in the planned breaking changes doc?

**Comment by pfrazee** [comment]
Yeah, I'll try to hit this soon

**Comment by pfrazee** [comment]
Ok, just had a moment to make the requested change. Done

File: atom/app/atom_content_client.h [hunk 3]
```diff
@@ -31,6 +31,9 @@ class AtomContentClient : public brightray::ContentClient {
       std::vector<content::PepperPluginInfo>* plugins) override;
   void AddServiceWorkerSchemes(
       std::set<std::string>* service_worker_schemes) override;
+  void AddSecureSchemesAndOrigins(
```
Comment (kevinsawicki) [inline atom/app/atom_content_client.h]: Is this still used? I'm not seeing references to it.

File: atom/renderer/api/atom_api_web_frame.cc [hunk 4]
```diff
@@ -4,6 +4,8 @@
 
 #include "atom/renderer/api/atom_api_web_frame.h"
 
+#include <iostream>
```
Comment (kevinsawicki) [inline atom/renderer/api/atom_api_web_frame.cc]: :fire:

Comment (pfrazee) [inline atom/renderer/api/atom_api_web_frame.cc]: removed

File: atom/app/atom_content_client.h (hunk 3 above)
Comment (pfrazee) [inline atom/app/atom_content_client.h]: I believe it's called by chromium or brightray code. I confirmed by removing it and running tests, which then failed.

Comment (kevinsawicki) [inline atom/app/atom_content_client.h]: Sorry, I didn't see the `override` 👍

File: atom/renderer/atom_renderer_client.cc [hunk 5]
```diff
@@ -121,20 +121,24 @@ bool IsDevToolsExtension(content::RenderFrame* render_frame) {
       .SchemeIs("chrome-extension");
 }
 
+std::vector<std::string> ParseSchemesCLISwitch(const char* switch_name) {
+  base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  std::string custom_schemes = command_line->GetSwitchValueASCII(switch_name);
+  if (!custom_schemes.empty()) {
```
Comment (kevinsawicki) [inline atom/renderer/atom_renderer_client.cc]: `base::SplitString` should handle empty strings so I don't think we need this check here.

File: atom/app/atom_content_client.cc [hunk 6]
```diff
@@ -204,4 +204,18 @@ void AtomContentClient::AddServiceWorkerSchemes(
   service_worker_schemes->insert(url::kFileScheme);
 }
 
+void AtomContentClient::AddSecureSchemesAndOrigins(
+    std::set<std::string>* secure_schemes,
+    std::set<GURL>* secure_origins) {
+  std::vector<std::string> schemes;
+  ConvertStringWithSeparatorToVector(&schemes, ",",
+                                     switches::kSecureSchemes);                        
+  if (!schemes.empty()) {
```
Comment (kevinsawicki) [inline atom/app/atom_content_client.cc]: I think this check would be okay to remove since the for loop should handle an empty schemes vector fine.

**Comment by kevinsawicki** [comment]
Fixed a few linter warnings and rebased to resolve the `planned-breaking-changes.md` conflict.

Thanks @pfrazee for this 👍

**Pull request merged**